### PR TITLE
types(v2): Convert sitemap plugin to TypeScript

### PR DESCRIPTION
**Pull request by moozzyk**
Test - enabled the sitemap plugin in the v2 website and verified that
the sitemap is created after running `docusaurus build`.

## Motivation

TypeScript offers strong typing.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Updated existing tests and made sure they pass.
Enabled the sitemap plugin in the v2 website and verified that the sitemap is created after running `docusaurus build`.

## Related PRs

No functional changes.

### CODE REVIEW

File: packages/docusaurus-plugin-sitemap/src/createSitemap.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import sitemap, {SitemapItemOptions} from 'sitemap';
+import {PluginOptions} from './types';
+import {DocusaurusConfig} from '@docusaurus/types';
+
+export default function createSitemap(
+  siteConfig: DocusaurusConfig,
```
Comment (wgao19) [inline packages/docusaurus-plugin-sitemap/src/createSitemap.ts]: any reason we change this function signature? any existing usages affected?

File: packages/docusaurus-plugin-sitemap/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+export interface PluginOptions {
```
Comment (wgao19) [inline packages/docusaurus-plugin-sitemap/src/types.ts]: any reason to use interface not type here?

File: packages/docusaurus-plugin-sitemap/src/createSitemap.ts (hunk 1 above)
Comment (moozzyk) [reply packages/docusaurus-plugin-sitemap/src/createSitemap.ts]: No one is affected - this is internal for the plugin. Without changing the signature I would have to introduce an ad hoc type for the passed object which looks quite ugly.

File: packages/docusaurus-plugin-sitemap/src/types.ts (hunk 2 above)
Comment (moozzyk) [reply packages/docusaurus-plugin-sitemap/src/types.ts]: Mostly for consistency with other plugins - .e.g. : https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-plugin-content-pages/src/types.ts or https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-plugin-content-blog/src/types.ts

**Comment by moozzyk** [comment]
I think I need exclude the `./lib` folder from linting somehow to make the CI happy.

File: packages/docusaurus-plugin-sitemap/src/__tests__/createSitemap.test.ts [hunk 3]
```diff
@@ -9,20 +9,21 @@ import createSitemap from '../createSitemap';
 
 describe('createSitemap', () => {
   test('simple site', () => {
-    const sitemap = createSitemap({
-      siteConfig: {
+    const sitemap = createSitemap(
```
Comment (endiliey) [inline packages/docusaurus-plugin-sitemap/src/__tests__/createSitemap.test.ts]: Make the typing stronger

```ts
    const sitemap = createSitemap(
      {
        url: 'https://example.com',
      } as DocusaurusConfig,
      ['/', '/test'],
      DEFAULT_OPTIONS,
    );
```

File: packages/docusaurus-plugin-sitemap/src/types.ts (hunk 2 above)
Comment (wgao19) [reply packages/docusaurus-plugin-sitemap/src/types.ts]: would need to bring to attention @yangshun @endiliey

**Comment by endiliey** [comment]
Please update changelog-2.x.md

File: packages/docusaurus-plugin-sitemap/src/types.ts (hunk 2 above)
Comment (endiliey) [reply packages/docusaurus-plugin-sitemap/src/types.ts]: this is okay. interface makes it easier to extend and even using advanced types like Partial

**Pull request merged**